keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -0,0 +1,477 @@
+
+### Goals
+
+Optionally (chosen by user), do not recursively relabel content of the volumes.
+
+### Non-Goals
+
+Change container runtimes / CRI.
+
+## Proposal
+
+Offer option in `Pod.Spec.PodSecurityContext to` *mount* volumes with the right labels instead of recursive relabeling:
+
+```go
+type SELinuxRelabelPolicy string
+
+const (
+    OnVolumeMount SELinuxRelabelPolicy = "OnVolumeMount"
+    AlwaysRelabel SELinuxRelabelPolicy = "Always"
+)
+
+type PodSecurityContext struct {
+    // seLinuxRelabelPolicy ← new field
+    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
+    // Valid values are "OnVolumeMount" and "Always". If not specified, "Always" is used.
+    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
+    // "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+    // change.
+    // This field is ignored for Pod's volumes that do not support SELinux.
+    // + optional
+    SELinuxRelabelPolicy *SELinuxRelabelPolicy
+
+    // For context:
+    // fsGroupChangePolicy defines behavior of changing ownership and permission of the volume
+    // before being exposed inside Pod. This field will only apply to
+    // volume types which support fsGroup based ownership(and permissions).
+    // It will have no effect on ephemeral volume types such as: secret, configmaps
+    // and emptydir.
+    // Valid values are "OnRootMismatch" and "Always". If not specified defaults to "Always".
+    // +optional
+    FSGroupChangePolicy *PodFSGroupChangePolicy `json:"fsGroupChangePolicy,omitempty" protobuf:"bytes,9,opt,name=fsGroupChangePolicy"`
+
+    ...
+}
+```
+
+See https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20200120-skip-permission-change.md for similar API for ownership change for fsGroup.
+This KEP should follow API provided for fsGroup closely, however, the implementation is different (`mount` here vs. recursive `chown` in the other KEP).
+
+In order to allow `SELinuxRelabelPolicy: OnVolumeMount` for volumes provided by CSI drivers, kubelet must know if a CSI driver supports SELinux or not.
+
+```go
+// In storage.k8s.io/v1:
+
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // SELinuxMountSupported specifies if the CSI driver supports "-o context"
+    // mount option.
+    //
+    // When "true", Kubernetes may call NodeStage / NodePublish with "-o context=xyz" mount
+    // option for volumes of a pod with
+    // podSecurityContext.seLinuxRelabelPolicy ="OnVolumeMount".
+    //
+    // When "false", Kubernetes won't pass any special SELinux mount options to the driver.
+    // podSecurityContext.seLinuxRelabelPolicy "OnVolumeMount" is silently ignored.
+    //
+    // Default is "false".
+    SELinuxMountSupporteded *bool;
```

Comment:
What about the case where the driver doesn't support it and we don't want to recursively relabel?

Reply:
This is currently autodetected by presence of `seclabel` mount option. If it's there after NodePublish, the mounted volume supports SELinux and CRI relabels, even in Kubernetes 1.18 and earlier.

This behavior will be the same with this KEP. I haven't heard requests not to relabel when `seclabel` is present.